File: src/reuser/agent/reuser.php
<?php
/*
 SPDX-FileCopyrightText: © 2014-2018 Siemens AG
Author: Daniele Fognini, Andreas Würl

 SPDX-License-Identifier: GPL-2.0-only
*/
/**
 * @dir
 * @brief Source for reuser agent
 * @page reuser Reuser agent
 * @tableofcontents
 * @section reuserabout About Reuser agent
 * If you have already uploaded a version of a package and clearing has been
 * done on it. While uploading the next version of it, you can reuse the
 * clearing decisions from the previous upload saving you lot of time.
 *
 * The files from both uploads (new and old) are compared using their hash.
 * If there is a match, the clearing decisions from reused upload file are
 * copied to the new upload file.
 *
 * Reuser also supports enhanced reuse mode (which can be slow). This mode
 * will copy a clearing decision if the two files differ by one line determined
 * by the `diff` tool.
 *
 * Reuser also supports reusing the main license(s) which copies any main
 * license decisions from old upload to new upload.
 *
 * @section reuseractions Supported actions
 * Currently, unified report agent does not support CLI commands and read only
 * from scheduler.
 *
 * @section reusersource Agent source
 *   - @link src/reuser/agent @endlink
 *   - @link src/reuser/ui @endlink
 *   - Functional test cases @link src/reuser/agent_tests/Functional @endlink
 */
/**
 * @namespace Fossology\Reuser
 * @brief Namespace used by reuser agent
 */
namespace Fossology\Reuser;

include_once(__DIR__."/ReuserAgent.php");

$agent = new ReuserAgent();
$agent->scheduler_connect();
$agent->run_scheduler_event_loop();
$agent->scheduler_disconnect(0);

File: src/reuser/agent/reuser.cc
/*
 SPDX-License-Identifier: GPL-2.0-only
 Author: Dietmar Helmut Leher <helmut.leher.ext@vaillant-group.com>
 SPDX-FileCopyrightText: © 2026 Vaillant GmbH
*/
/**
 * @dir
 * @brief The reuser agent
 * @file
 * @brief Entry point for the reuser agent
 * @page reuser Reuser Agent
 * @tableofcontents
 *
 * The reuser agent copies clearing decisions, copyright events, main
 * licenses and report configuration from a previously cleared upload to a
 * new upload, based on pfile identity (standard reuse), file-name matching
 * with diff-threshold (enhanced reuse), or both.
 *
 * @section reuseractions Supported modes (set in upload_reuse.reuse_mode)
 * | Bit | Constant         | Description                          |
 * | --: | :--------------- | :----------------------------------- |
 * |   2 | REUSE_ENHANCED   | Match by filename + diff threshold   |
 * |   4 | REUSE_MAIN       | Copy main-license entries            |
 * |  16 | REUSE_CONF       | Copy report configuration            |
 * | 128 | REUSE_COPYRIGHT  | Copy copyright events                |
 *
 * @section reusersource Agent source
 *   - @link src/reuser/agent @endlink
 */

#include "ReuserUtils.hpp"

int main(int argc, char** argv)
{
  fo::DbManager dbManager(&argc, argv);
  ReuserDatabaseHandler databaseHandler(dbManager);

  ReuserState state = getState(dbManager);

  while (fo_scheduler_next() != nullptr)
  {
    int uploadId = atoi(fo_scheduler_current());
    if (uploadId == 0) continue;

    int arsId = writeARS(state, 0, uploadId, 0, dbManager);
    if (arsId <= 0)
      bail(5);

    if (!processUploadId(state, uploadId, databaseHandler))
      bail(2);

    fo_scheduler_heart(0);
    writeARS(state, arsId, uploadId, 1, dbManager);
  }

  fo_scheduler_heart(0);
  fo_scheduler_disconnect(0);
  return 0;
}
